### DISCUSSION TITLE
Precedence doesn’t change.

### DISCUSSION THREAD

**Discussion by eugenelepekhin** [antlr/antlr4#4175] [Q&A]
Hi all,
Consider the following snippet of grammar:
```
expr: If Open cond=expr Close Begin trueBranch=exprList End (Else Begin falseBranch=exprList End)? #If
	| {IsMacroN()}? macroName arguments	#Call
	| {IsMacro0()}? macroName			#Call
	| Open expr Close					#ParenExpr
	| (Add | Not) expr					#Unary
	| left=expr op=Mul right=expr		#Bin
	| left=expr op=Add right=expr		#Bin
…
	| {IsNotMacro()}? name				#LocalName
	| NumberLiteral						#Literal
	| StringLiteral						#Literal
	;

```
The expression:
`dub 2 + 3
`
get parsed like this:
`(dub (2 + 3))
`
So, it’s first flavor of #Call. The predicate IsMacroN() is returning true for that expression.
If I move both calls down, right above the #LocalName branch the expression get parsed in exactly the same tree while I’d expect it to be:
`((dub 2) + 3)
`
Actually, I’d probably expect is the other way around in the above state of the grammar to be the second parse tree, but you probably get the point.
Can someone please explain why moving #Call does not change the precedence?

**Comment by ericvergnaud** [comment]
Hi,
Your rule has duplicate labels so should not generate a parser. 
Maybe check the output of the build ?

**Reply by eugenelepekhin** [reply]
It actually generates the parser and work fine with duplicate labels.

**Reply by ericvergnaud** [reply]
May I suggest that you triple check ?

**Reply by eugenelepekhin** [reply]
Sure, this is very nice feature and here is the documentation on it:
https://github.com/antlr/antlr4/blob/master/doc/parser-rules.md#alternative-labels

**Reply by ericvergnaud** [reply]
Fair enough, I wasn't aware of that 'feature'. I'd be surprised if it doesn't affect precedence. Have you tried using distinct labels ? Also worth mentioning that it's possible that semantic predicates affect precedence

**Reply by ericvergnaud** [reply]
removing labels and predicates triggers the same behavior, which is believe is expected:
- first match for 'dub' is 'macroName' in 'macroName arguments'
- next, antlr checks is there's a match for arguments (which I guess is expr+ ?) and finds 'exp Add exp'

In other words, the precedence issue is between NumberLiteral and 'exp Add exp', but swapping them doesn't help either because antlr tries to consume as many tokens it can.

not sure you can get the expected behavior with your grammar.  If you want function calls to only consume 1 argument (to form monads) then you need to move binary expressions out of expr rule I think

**Reply by eugenelepekhin** [reply]
Yes, that is probably the reason. Thanks.

**Reply by n-dragon** [reply]
> removing labels and predicates triggers the same behavior, which is believe is expected:
> 
> * first match for 'dub' is 'macroName' in 'macroName arguments'
> * next, antlr checks is there's a match for arguments (which I guess is expr+ ?) and finds 'exp Add exp'
> 
> In other words, the precedence issue is between NumberLiteral and 'exp Add exp', but swapping them doesn't help either because antlr tries to consume as many tokens it can.
> 
> not sure you can get the expected behavior with your grammar. If you want function calls to only consume 1 argument (to form monads) then you need to move binary expressions out of expr rule I think

Could you elaborate and the many tokens consumption behavior ?
this would mean precedence only applies on same token length alternatives ?
